### PR TITLE
Closes #140

### LINKED ISSUE #140: Create a dataset loader for TwADR-L
- Task: NER, NEN
- License: Creative Commons Attribution 4.0 International
- Format: BioC
- Language: English
- Citation:
@inproceedings{limsopatham2016normalising,
  title={Normalising medical concepts in social media texts by learning semantic representation},
  author={Limsopatham, Nut and Collier, Nigel},
  booktitle={Proceedings of the 54th annual meeting of the association for computational linguistics (volume 1: long papers)},
  pages={1014--1023},
  year={2016}
}
- Source: https://zenodo.org/record/55013#.YgpAHVvMJhE

### PR DESCRIPTION
**Pull request by j-chim**
Please name your PR after the issue it closes. You can use the following line: "Closes #ISSUE-NUMBER" where you replace the ISSUE-NUMBER with the one corresponding to your dataset.

If the following information is NOT present in the issue, please populate:

- **Name:** *name of the dataset*
- **Description:** *short description of the dataset (or link to social media or blog post)*
- **Paper:** *link to the dataset paper if available*
- **Data:** *link to the online home of the dataset*

### Checkbox

- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `biodatasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_BIGBIO_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `BigBioConfig` for the source schema and one for a bigbio schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_bigbio biodatasets/my_dataset/my_dataset.py`.

### CODE REVIEW

File: biodatasets/twadrl/twadrl.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+            subset_id="twadrl",
+        ),
+    ]
+
+    def _info(self):
+        if self.config.schema == "source":
+            features = datasets.Features(
+                {
+                    "fold": datasets.Value("int32"),
+                    "cui": datasets.Value("string"),
```
Comment (sg-wbi) [inline biodatasets/twadrl/twadrl.py:88]: minor, but wouldn't an `int` be better?

File: biodatasets/twadrl/twadrl.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            ]
+        ]
+
+    def _generate_examples(self, filepaths, split):
+        for filepath in filepaths:
+            fold_id = re.search("TwADR-L\.fold-(\d)\.", filepath).group(1)
+            with open(filepath, "r", encoding="latin-1") as f:
+                document_id = f"{split}_{fold_id}"
+                for i, line in enumerate(f):
+                    id = f"{document_id}_{fold_id}_{i}"
```
Comment (sg-wbi) [inline biodatasets/twadrl/twadrl.py:130]: This is a good way to preserve the fold information. I am not sure what the consensus is on the k-fold datasets. I;ll ask around

**Review by sg-wbi — COMMENTED** [review]
Thank you for your contribution! Everything looks fine to me. The only issue is with the k-fold splits which I am not sure how we want to handle. I will ask and maybe you can too on the `#question` in Discord?  This way we get more visibility

File: biodatasets/twadrl/twadrl.py (hunk 1 above)
Comment (j-chim) [reply biodatasets/twadrl/twadrl.py:88]: Just to double check, "int" instead of "int32" for `fold`? Or for `cui`? 

If the former, I'm not familiar with how integer size is handled in datasets/pyarrow (I'm guessing it'll default to 32?) so thought I should specify it. Maybe "int8" is more appropriate for number of folds? Would be happy to change this.

If the latter, the CUIs are strings here are always prefixed with "C".

File: biodatasets/twadrl/twadrl.py (hunk 2 above)
Comment (j-chim) [reply biodatasets/twadrl/twadrl.py]: Based on Discord, have updated dataloader so that each of the k folds is its own split (eg "train_0", "train_1", ..., "train_k")

File: biodatasets/twadrl/twadrl.py (hunk 1 above)
Comment (sg-wbi) [reply biodatasets/twadrl/twadrl.py:87]: > If the latter, the CUIs are strings here are always prefixed with "C".

Argh! Sorry I got it mixed up with the other dataset. for "cui" string is fine. For "fold" "int32" is fine. Sorry for the confusion.

File: biodatasets/twadrl/twadrl.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+                            }
+                        ],
+                        "entities": [ 
+                            {
+                                "id": f"{id}_entity",
+                                "type": text_type,
+                                "text": [social_media_text],
+                                "offsets" : [offset],
+                                "normalized" : [
+                                    {"db_name" : "SNOMED-CT|AMT", "db_id" : cui}
```
Comment (sg-wbi) [inline biodatasets/twadrl/twadrl.py:160]: Could you change `db_name` to `SIDER 4` please?
According to the paper:
> This collection contains 1,436 Twitter phrases that can be mapped to one of 2,220 medical concepts from the SIDER 4 database of drug profiles

File: biodatasets/twadrl/twadrl.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                                    ]
+                            }
+                        ],
+                        "events": [],
+                        "coreferences": [],
+                        "relations": [],
+                    }
```
Comment (sg-wbi) [inline biodatasets/twadrl/twadrl.py:167]: Could you please fix this?

**Review by sg-wbi — COMMENTED** [review]
Thank you for handling the k-fold thing! Could you please address these two minor changes? Then it should be good to go.

File: biodatasets/twadrl/twadrl.py (hunk 4 above)
Comment (sg-wbi) [reply biodatasets/twadrl/twadrl.py:167]: Missing newline at the end of the file

File: biodatasets/twadrl/twadrl.py (hunk 3 above)
Comment (j-chim) [reply biodatasets/twadrl/twadrl.py]: Thanks for catching this, have corrected it here & in the description in the latest commit and have fixed the formatting.

**Comment by sg-wbi** [comment]
Thank you for fixing this! One last (for real) very minor thing:
```python
{'id': 'train_0_3603', 'document_id': 'train_0', 'passages': [{'id': 'train_0_3603_passage',
```
In the bigbio schema the `document_id` is always the same for all examples (`split_kfold-idx`). Could you please change this to make it unique? I guess it's going to be the same for #99

**Comment by j-chim** [comment]
> Thank you for fixing this! One last (for real) very minor thing:
> 
> ```python
> {'id': 'train_0_3603', 'document_id': 'train_0', 'passages': [{'id': 'train_0_3603_passage',
> ```
> 
> In the bigbio schema the `document_id` is always the same for all examples (`split_kfold-idx`). Could you please change this to make it unique? I guess it's going to be the same for #99

Got it! I think it should be fixed now. For example, the `document_id` for the first fold in train should now look like `"train_kfold-0"`

**Comment by sg-wbi** [comment]
I am sorry for the confusion but I actually meant:
```python
{'id': 'train_0_3603', 'document_id': 'train_0_3603', 'passages': [{'id': 'train_0_3603_passage'
```
But I guess it does not matter that much. I'm just thinking of the scenario of searching the dataset by `document_id`. I would expect to get a single item back. What do you think?

**Comment by j-chim** [comment]
> I am sorry for the confusion but I actually meant:
> 
> ```python
> {'id': 'train_0_3603', 'document_id': 'train_0_3603', 'passages': [{'id': 'train_0_3603_passage'
> ```
> 
> But I guess it does not matter that much. I'm just thinking of the scenario of searching the dataset by `document_id`. I would expect to get a single item back. What do you think?

Ah I see what you mean now. 

The reason I went with ‘train_0’ initially was that it corresponds directly to the original file. In that case searching document_id for ‘train_0’ will give everything in fold 0. 

But now that each fold is its own split this is no longer necessary. I think your format makes more sense. Will change to that
